pkg/controllers/managementuser/rbac/roletemplates/prtb_handler.go

```diff
@@ -0,0 +1,253 @@
+}
+
+// getNamespacesFromProject Lists all namespaces within a project
+func (p *prtbHandler) getNamespacesFromProject(prtb *v3.ProjectRoleTemplateBinding) (*corev1.NamespaceList, error) {
+	_, projectId, _ := strings.Cut(prtb.ProjectName, ":")
+	return p.nsClient.List(metav1.ListOptions{
+		LabelSelector: projectIDAnnotation + "=" + projectId,
+	})
+}
+
+// buildRoleBinding creates a role binding owned by the prtb.
+func buildRoleBinding(prtb *v3.ProjectRoleTemplateBinding, roleRefName string, subject rbacv1.Subject) *rbacv1.RoleBinding {
+	ownerLabel := createPRTBOwnerLabel(prtb.Name)
+	roleRef := rbacv1.RoleRef{
+		Kind: "Role",
+		Name: roleRefName,
+	}
+	return &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "rb-",
+			Labels:       map[string]string{ownerLabel: "true"},
+		},
+		RoleRef:  roleRef,
+		Subjects: []rbacv1.Subject{subject},
+	}
+}
+
+// ensureOnlyDesiredRoleBindingsExist finds any RoleBindings owned by the PRTB, and removes them if they don't match either of our desired RoleBindings.
+// If our desired RoleBindings aren't found, it creates them.
+// If desiredPromotedRB is left nil, only check for desiredRB.
+func (p *prtbHandler) ensureOnlyDesiredRoleBindingsExist(desiredRB, desiredPromotedRB *rbacv1.RoleBinding, namespace, prtbOwnerLabel string) error {
+	// Check if any Role Bindings exist already
+	currentRBs, err := p.rbClient.List(namespace, metav1.ListOptions{LabelSelector: prtbOwnerLabel})
+	if err != nil || currentRBs == nil {
+		return err
+	}
+
+	var matchingRB, matchingPromotedRB *rbacv1.RoleBinding
+	var rbsToDelete []*rbacv1.RoleBinding
+	// Search for RoleBindings that are needed, all others should be removed
+	for _, currentRB := range currentRBs.Items {
+		if areRoleBindingsSame(&currentRB, desiredRB) && matchingRB == nil {
+			matchingRB = &currentRB
+		} else if desiredPromotedRB != nil && areRoleBindingsSame(&currentRB, desiredPromotedRB) && matchingPromotedRB == nil {
+			matchingPromotedRB = &currentRB
+		} else {
+			rbsToDelete = append(rbsToDelete, &currentRB)
+		}
+	}
+
+	// Remove excess RBs
+	for _, rbToDelete := range rbsToDelete {
+		if err = p.rbClient.Delete(namespace, rbToDelete.Name, &metav1.DeleteOptions{}); err != nil {
+			return err
+		}
+	}
+
+	if matchingRB == nil {
+		if _, err := p.rbClient.Create(desiredRB); err != nil {
+			return err
+		}
+	}
+	if desiredPromotedRB != nil && matchingPromotedRB == nil {
+		if _, err := p.rbClient.Create(desiredPromotedRB); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// areRoleBindingsSame compares the Subjects and RoleRef fields of two Role Bindings.
+func areRoleBindingsSame(rb1, rb2 *rbacv1.RoleBinding) bool {
+	return reflect.DeepEqual(rb1.Subjects, rb2.Subjects) &&
+		reflect.DeepEqual(rb1.RoleRef, rb2.RoleRef)
+}
+
+// createPRTBOwnerLabel creates an owner label given a PRTB name.
+func createPRTBOwnerLabel(prtbName string) string {
+	return name.SafeConcatName(prtbOwnerLabel, prtbName)
+}
```

Comment:
Doesn't this produce something like`authz.cluster.cattle.io/prtb-owner-$prtbName=true`, making it harder to use this as a label? (also increasing the cardinality of labels, but I don't know if that's a problem or not). Why not using `authz.cluster.cattle.io/prtb-owner=$prtbName` instead?